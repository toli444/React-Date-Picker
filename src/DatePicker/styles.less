.date-picker {
    margin-top: 30vh;
}

.calendar-heading {
    display: flex;
}
