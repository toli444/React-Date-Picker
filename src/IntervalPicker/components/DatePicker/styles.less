.date-picker {
  display: inline-block;
  outline: 0;
  margin-right: 3px;

  .selected-date-container {
    display: flex;
    flex-direction: column;
    outline: 0;

    > label {
      margin-bottom: 0;
    }

    .selected-date {
      height: 33px;
    }
  }

  &:focus-within {
    outline: 3px solid rgba(29, 144, 255, 0.75);
    background-color: rgba(29, 144, 255, 0.75);

    .selected-date {
      z-index: 1;
    }

    .calendar {
      outline: 3px solid rgba(29, 144, 255, 0.75);
    }
  }

  .calendar {
    position: absolute;
    background-color: #fefef4;
    width: 300px;

    .heading {
      display: flex;
      width: 100%;
    }
  }
}
