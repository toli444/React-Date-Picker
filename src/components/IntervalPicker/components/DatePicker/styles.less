.date-picker {
  display: inline-block;
  outline: 0;
  margin-right: 3px;
  position: relative;

  .selected-date-container {
    display: flex;
    flex-direction: column;
    outline: 0;

    > label {
      margin-bottom: 0;
    }

    .selected-date {
      height: 33px;
    }
  }

  &:focus-within {
    .selected-date-container {
      outline: 3px solid rgba(29, 144, 255, 0.75);
      background-color: rgba(29, 144, 255, 0.75);
    }

    .selected-date {
      z-index: 2;
    }

    .calendar {
      outline: 3px solid rgba(29, 144, 255, 0.75);
    }
  }

  &:global(.right) {
    .calendar {
      right: 0;
    }
  }

  .calendar {
    position: absolute;
    background-color: #fefef4;
    width: 300px;
    z-index: 1;

    .heading {
      display: flex;
      width: 100%;
    }
  }
}
