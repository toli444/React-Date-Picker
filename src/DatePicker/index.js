import React, {Component} from 'react';
import moment from 'moment';
import styles from './styles.less';

import SingleArrow from './components/SingleArrow';
import DoubleArrow from './components/DoubleArrow';

class DatePicker extends Component {
  static defaultProps = {
    date: '11-11-2011',
    format: 'DD-MM-YYYY',
  };

  constructor(props) {
    super(props);

    this.state = {
      selectedDate: props.date ? moment(props.date).format(props.format) : '',
      isCalendarShown: false,
    };
  }

  subDate = () => {
  };

  toggleCalendarShown = (isCalendarShown) => {
    this.setState({isCalendarShown});
  };

  render() {
    const {selectedDate, isCalendarShown} = this.state;

    return (
      <div className={styles['date-picker']}>
        <div>
          <input
            type="text"
            className="selected-date"
            value={selectedDate}
            onFocus={() => this.toggleCalendarShown(true)}
            onBlur={() => this.toggleCalendarShown(false)}
          />
        </div>
        {isCalendarShown && <div>
          <div className={styles['calendar-heading']}>
            <DoubleArrow
              left
              onClick={() => this.subDate(1, 'year')}
            />
            <SingleArrow
              left
              onClick={() => this.subDate(1, 'month')}
            />
            <input className="pick-month-year" value="ttt"/>
            <SingleArrow onClick={() => this.addDate(1, 'month')}/>
            <DoubleArrow onClick={() => this.addDate(1, 'year')}/>
          </div>
          {/*<CalendarBody*/}
          {/*date={dateToEdit}*/}
          {/*format={format}*/}
          {/*maxDate={maxDate}*/}
          {/*minDate={minDate}*/}
          {/*onChange={(value, unit) => this.handleDateChange(value, unit)}*/}
          {/*viewFor={setViewFor}*/}
          {/*yearBlock={yearBlock}*/}
          {/*/>*/}
        </div>
        }
      </div>
    );
  }
}

export default DatePicker;
