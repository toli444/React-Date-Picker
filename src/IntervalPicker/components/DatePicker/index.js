import React, {Component} from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';

import styles from './styles.less';

import EditingDateInput from '../EditingDateInput';
import Arrow from '../Arrow';
import DaysChooser from '../choosers/DaysChooser';

class DatePicker extends Component {
  static propTypes = {
    className: PropTypes.string,
    value: PropTypes.object,
    startDate: PropTypes.object,
    endDate: PropTypes.object,
    minDate: PropTypes.object,
    maxDate: PropTypes.object,
    format: PropTypes.string,
    onDateChange: PropTypes.func,
    label: PropTypes.string,
  };

  static defaultProps = {
    className: ''
  };

  blurTimeout = null;

  constructor(props) {
    super(props);

    this.containerRef = React.createRef();
    this.state = {
      currentDate: this.getCurrentDate(),
      isFocusIn: null,
      editingUnit: null,
    };
  }

  getCurrentDate() {
    const {value, startDate, format} = this.props;
    const actualValue = value || startDate;

    return actualValue ? moment(actualValue, format) : null
  }

  addDate(amount, unit) {
    this.setState({
      currentDate: moment(this.state.currentDate, this.props.format).add(amount, unit),
      editingUnit: unit
    })
  }

  subDate(amount, unit) {
    this.setState({
      currentDate: moment(this.state.currentDate, this.props.format).subtract(amount, unit),
      editingUnit: unit
    })
  }

  focusIn() {
    this.containerRef.current.focus();
  }

  onBlurHandler = () => {
    this.blurTimeout = setTimeout(() => this.setState({ isFocusIn: false }));
  };

  onFocusHandler = () => {
    clearTimeout(this.blurTimeout);

    if (!this.state.isFocusIn) {
      this.setState({
        isFocusIn: true,
        currentDate: this.getCurrentDate(),
      });
    }
  };

  handleDateChange = newDate => {
    this.setState({currentDate: newDate, isFocusIn: false});
    this.props.onDateChange(newDate);
  };

  render() {
    const {format, value, startDate, endDate, minDate, maxDate, label, className} = this.props;
    const {currentDate, isFocusIn, editingUnit} = this.state;

    return (
      <div
        className={`${styles['date-picker']} ${className}`}
        onFocus={this.onFocusHandler}
        onBlur={this.onBlurHandler}
        tabIndex="0"
        ref={this.containerRef}
      >
        <div className={styles['selected-date-container']}>
          <label>{label}</label>
          <input
            type="text"
            className={styles['selected-date']}
            value={value ? value.format(format) : ''}
            readOnly
          />
        </div>
        {isFocusIn && <div className={styles.calendar}>
          <div className={styles.heading}>
            <Arrow left double onClick={() => this.subDate(1, 'year')}/>
            <Arrow left onClick={() => this.subDate(1, 'month')}/>
            <EditingDateInput currentDate={currentDate} editingUnit={editingUnit} format={format}/>
            <Arrow onClick={() => this.addDate(1, 'month')}/>
            <Arrow double onClick={() => this.addDate(1, 'year')}/>
          </div>
          {<DaysChooser
            currentDate={currentDate}
            startDate={startDate}
            endDate={endDate}
            onDateChange={this.handleDateChange}
            minDate={minDate}
            maxDate={maxDate}
          />}
        </div>
        }
      </div>
    );
  }
}

export default DatePicker;
