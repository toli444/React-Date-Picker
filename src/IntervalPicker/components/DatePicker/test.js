import React from 'react';
import DatePicker from '../DatePicker';
import { shallow, mount } from 'enzyme';

import moment from "moment";

const format = 'DD/MM/YYYY';

describe('DatePicker', () => {
  it('renders correctly', () => {
    const wrapper  = shallow(<DatePicker />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with focusIn', () => {
    const wrapper  = shallow(<DatePicker />);
    wrapper.instance().onFocusHandler();
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('getCurrentDate', () => {
    const startDate = moment('28/07/1997', format);
    const value = moment('29/07/1997', format);
    const wrapper  = shallow(<DatePicker
      format={format}
      startDate={startDate}
      value={value}
    />);
    const instance = wrapper.instance();

    expect(instance.getCurrentDate()).toEqual(value);
    wrapper.setProps({ startDate: null, value});
    expect(instance.getCurrentDate()).toEqual(value);
    wrapper.setProps({ startDate, value: null});
    expect(instance.getCurrentDate()).toEqual(startDate);
    wrapper.setProps({ startDate: null, value: null});
    expect(instance.getCurrentDate()).toEqual(null);
    wrapper.setProps({ startDate: undefined, value: undefined});
    expect(instance.getCurrentDate()).toEqual(null);
  });

  it('addDate', () => {
    const wrapper  = shallow(<DatePicker value={moment('28/07/1997', format)} />);
    const instance = wrapper.instance();

    expect(instance.state.currentDate.format()).toEqual("1997-07-28T00:00:00+03:00");
    instance.addDate(1, 'month');
    expect(instance.state.currentDate.format()).toEqual("1997-08-28T00:00:00+03:00");
    instance.addDate(1, 'year');
    expect(instance.state.currentDate.format()).toEqual("1998-08-28T00:00:00+03:00");
    instance.addDate(10, 'year');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.addDate(0, 'year');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.addDate();
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.addDate(1, 'blabla');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
  });

  it('subDate', () => {
    const wrapper  = shallow(<DatePicker value={moment('28/08/2008', format)} />);
    const instance = wrapper.instance();

    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.subDate(1, 'blabla');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.subDate();
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.subDate(0, 'year');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.subDate(10, 'year');
    expect(instance.state.currentDate.format()).toEqual("1998-08-28T00:00:00+03:00");
    instance.subDate(1, 'year');
    expect(instance.state.currentDate.format()).toEqual("1997-08-28T00:00:00+03:00");
    instance.subDate(1, 'month');
    expect(instance.state.currentDate.format()).toEqual("1997-07-28T00:00:00+03:00");
  });

  it('onBlurHandler', () => {
    jest.useFakeTimers();
    const wrapper  = mount(<DatePicker />);
    const instance = wrapper.instance();

    expect(instance.blurTimeout).toEqual(null);
    expect(instance.state.isFocusIn).toEqual(null);
    wrapper.instance().onBlurHandler();
    jest.runAllTimers();
    expect(instance.blurTimeout).toEqual(1);
    expect(instance.state.isFocusIn).toEqual(false);
  });

  it('onFocusHandler', () => {
    jest.useFakeTimers();
    const wrapper  = mount(<DatePicker />);
    const instance = wrapper.instance();

    instance.getCurrentDate = jest.fn().mockReturnValue(moment('28/08/2008', format));
    instance.blurTimeout = setTimeout(() => {});

    expect(instance.blurTimeout).toEqual(1);
    expect(instance.state.isFocusIn).toEqual(null);
    expect(instance.state.currentDate).toEqual(null);
    wrapper.instance().onFocusHandler();
    expect(clearTimeout).toHaveBeenCalledWith(1);
    expect(instance.state.isFocusIn).toEqual(true);
    expect(instance.state.currentDate).toEqual(moment('28/08/2008', format));
  });
});
