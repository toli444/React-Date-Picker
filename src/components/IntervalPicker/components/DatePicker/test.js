import React from 'react';
import DatePicker from './index';
import { shallow, mount } from 'enzyme';

import moment from "moment";
import MockDate from "mockdate";

const format = 'DD/MM/YYYY';

describe('DatePicker', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    MockDate.set(moment('28/07/1997', format));
  });

  afterEach(() => {
    MockDate.reset();
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('renders correctly', () => {
    const wrapper  = shallow(<DatePicker />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with focusIn', () => {
    const wrapper  = shallow(<DatePicker />);
    wrapper.instance().onFocusHandler();
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('getCurrentDate', () => {
    const startDate = moment('28/07/1997', format);
    const value = moment('29/07/1997', format);
    const wrapper  = shallow(<DatePicker
      format={format}
      startDate={startDate}
      value={value}
    />);
    const instance = wrapper.instance();

    expect(instance.getCurrentDate()).toEqual(value);
    wrapper.setProps({ startDate: null, value});
    expect(instance.getCurrentDate()).toEqual(value);
    wrapper.setProps({ startDate, value: null});
    expect(instance.getCurrentDate()).toEqual(startDate);
    wrapper.setProps({ startDate: null, value: null});
    expect(instance.getCurrentDate()).toEqual(moment());
    wrapper.setProps({ startDate: undefined, value: undefined});
    expect(instance.getCurrentDate()).toEqual(moment());
  });

  it('addDate', () => {
    const wrapper  = shallow(<DatePicker value={moment('28/07/1997', format)} />);
    const instance = wrapper.instance();

    expect(instance.state.currentDate.format()).toEqual("1997-07-28T00:00:00+03:00");
    instance.addDate(1, 'month');
    expect(instance.state.currentDate.format()).toEqual("1997-08-28T00:00:00+03:00");
    instance.addDate(1, 'year');
    expect(instance.state.currentDate.format()).toEqual("1998-08-28T00:00:00+03:00");
    instance.addDate(10, 'year');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.addDate(0, 'year');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.addDate();
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.addDate(1, 'blabla');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
  });

  it('subDate', () => {
    const wrapper  = shallow(<DatePicker value={moment('28/08/2008', format)} />);
    const instance = wrapper.instance();

    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.subDate(1, 'blabla');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.subDate();
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.subDate(0, 'year');
    expect(instance.state.currentDate.format()).toEqual("2008-08-28T00:00:00+03:00");
    instance.subDate(10, 'year');
    expect(instance.state.currentDate.format()).toEqual("1998-08-28T00:00:00+03:00");
    instance.subDate(1, 'year');
    expect(instance.state.currentDate.format()).toEqual("1997-08-28T00:00:00+03:00");
    instance.subDate(1, 'month');
    expect(instance.state.currentDate.format()).toEqual("1997-07-28T00:00:00+03:00");
  });

  it('onBlurHandler', () => {
    const wrapper  = mount(<DatePicker />);
    const instance = wrapper.instance();

    expect(instance.blurTimeout).toEqual(null);
    expect(instance.state.isFocusIn).toEqual(null);
    instance.onBlurHandler();
    jest.runAllTimers();
    expect(instance.blurTimeout).not.toBe(false);
    expect(instance.state.isFocusIn).toEqual(false);
    jest.clearAllTimers()
  });

  it('onFocusHandler', () => {
    const wrapper  = mount(<DatePicker />);
    const instance = wrapper.instance();

    instance.getCurrentDate = jest.fn().mockReturnValue(moment('28/08/2008', format));
    instance.blurTimeout = setTimeout(() => {});

    expect(instance.blurTimeout).not.toBe(false);
    expect(instance.state.isFocusIn).toEqual(null);
    expect(instance.state.currentDate).toEqual(moment());
    instance.onFocusHandler();
    expect(clearTimeout).toHaveBeenCalledWith(instance.blurTimeout);
    expect(instance.state.isFocusIn).toEqual(true);
    expect(instance.state.currentDate).toEqual(moment('28/08/2008', format));
    jest.clearAllTimers()
  });

  it('handleDateChange', () => {
    const onDateChangeMock = jest.fn();
    const wrapper  = mount(<DatePicker
      onDateChange={onDateChangeMock}
    />);
    const instance = wrapper.instance();

    const execTestCase = (isFocusIn, currentDate, newDate, expectedFocusIn, expectedCurrentDate = newDate) => {
      expect(instance.state.isFocusIn).toEqual(isFocusIn);
      expect(instance.state.currentDate).toEqual(currentDate);
      instance.handleDateChange(newDate);
      expect(onDateChangeMock).toHaveBeenCalledWith(expectedCurrentDate);
      expect(instance.state.isFocusIn).toEqual(expectedFocusIn);
      expect(instance.state.currentDate).toEqual(expectedCurrentDate);
      jest.clearAllMocks();
    };

    execTestCase(null, moment(), moment('28/08/2008', format), false);
    execTestCase(false, moment('28/08/2008', format), null, false);
  });
});
