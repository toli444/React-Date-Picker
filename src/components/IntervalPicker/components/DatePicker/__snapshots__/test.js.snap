// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker renders correctly 1`] = `
<div
  className="date-picker "
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="0"
>
  <div
    className="selected-date-container"
  >
    <label />
    <input
      className="selected-date"
      readOnly={true}
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`DatePicker renders correctly with focusIn 1`] = `
<div
  className="date-picker "
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="0"
>
  <div
    className="selected-date-container"
  >
    <label />
    <input
      className="selected-date"
      readOnly={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="calendar"
  >
    <div
      className="heading"
    >
      <Arrow
        double={true}
        left={true}
        onClick={[Function]}
      />
      <Arrow
        left={true}
        onClick={[Function]}
      />
      <EditingDateInput
        currentDate={"1997-07-27T21:00:00.000Z"}
        editingUnit={null}
      />
      <Arrow
        onClick={[Function]}
      />
      <Arrow
        double={true}
        onClick={[Function]}
      />
    </div>
    <DaysChooser
      currentDate={"1997-07-27T21:00:00.000Z"}
      onDateChange={[Function]}
    />
  </div>
</div>
`;
